### PR TITLE
updateGolangSysHook: init

### PR DESCRIPTION
**Pull request by wegank**
###### Description of changes

~This PR adds a setup hook that updates `golang.org/x/sys` to a version that supports all Hydra platforms.~

Marked as a draft for its controversy. For all packages mentioned in this PR, anyone interested is encouraged to submit a patch upstream and then fetchPatch from the upstream PR.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by markus1189** [comment]
Can you give a bit more context about the patch?  It seems like it's only necessary for darwin according to the PR, but it's applied for all architectures?

**Comment by wegank** [comment]
@markus1189 Basically the same reason as in #184919: `golang.org/x/sys` is too old to build on aarch64-darwin.

The patch for Go 1.16 is always adding

```nix
golang.org/x/sys v0.0.0-20220731174439-a90be440212d // indirect
```

in `go.mod` and

```nix
golang.org/x/sys v0.0.0-20220731174439-a90be440212d h1:Sv5ogFZatcgIMMtBSTTAgMYsicp25MXBubjXNDKwm80=
golang.org/x/sys v0.0.0-20220731174439-a90be440212d/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
```

in `go.sum`.

However, the patch may not apply for Go 1.17, to which the upstream has migrated. I therefore apply the patch indifferently on all platforms to not cause the regression on updates.

**Comment by coderfromhere** [comment]
Isn't it easier to fix `x/sys` at https://github.com/instrumenta/kubeval/blob/master/go.sum for all Go packages that utilise `sys` in nixpkgs rather than copy-pasting the patch across multiple derivations?

**Comment by wegank** [comment]
@coderfromhere Done.

**Comment by coderfromhere** [comment]
@wegank yeah, I realised it wouldn't be as easy as I initially imagined and removed my comment, but huge thanks for actually wrapping it away into a self-contained thing!

**Comment by avanov** [comment]
Looks like the current `pkgs.kubeval` in master suffers from the same issue with `sys`. Would it be possible to add it into the same PR changelist?

File: pkgs/build-support/setup-hooks/update-golang-sys-hook.sh
Comment (SuperSandro2000) [inline pkgs/build-support/setup-hooks/update-golang-sys-hook.sh:7]:
```suggestion
    golang.org/x/sys v0.1.0 // indirect
```
lets use the recent 0.1.0 tag

**Comment by SuperSandro2000** [comment]
> but it's applied for all architectures?

Yes, to reduce maintenance. If possible patches should always be done on all platforms otherwise people on the other platform have no chance to do a basic verification if things could still work and we would have way more stale patches.

**Comment by wegank** [comment]
@avanov Added.

File: pkgs/build-support/setup-hooks/update-golang-sys-hook.sh
Comment (SuperSandro2000) [inline pkgs/build-support/setup-hooks/update-golang-sys-hook.sh:10]: Is this always working? go can be very nity about its vendoring files. What about modules.txt?

File: pkgs/build-support/setup-hooks/update-golang-sys-hook.sh
Comment (SuperSandro2000) [inline pkgs/build-support/setup-hooks/update-golang-sys-hook.sh:3]:
```suggestion
preConfigureHooks+=(updateGolangSys)

updateGolangSys() {
```

File: pkgs/build-support/setup-hooks/update-golang-sys-hook.sh
Comment (wegank) [reply pkgs/build-support/setup-hooks/update-golang-sys-hook.sh:7]: The problem is that 0.1.0 only supports Go 1.17+, while `awsls` was anchored to `go 1.16` in `go.mod`. Maybe we should rename this to `updateGolangSysHook116` or something?

File: pkgs/build-support/setup-hooks/update-golang-sys-hook.sh
Comment (wegank) [reply pkgs/build-support/setup-hooks/update-golang-sys-hook.sh:10]: Indeed, there are packages on which the hook doesn't work, e.g. `podgrab`, with the error

```
golang.org/x/sys@v0.0.0-20220731174439-a90be440212d: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
```

Otherwise, the hook already works pretty well on most failing packages on Hydra.

**Review by zowoq — CHANGES_REQUESTED** [review]
As one of the nixpkgs golang maintainers I very strongly object to this PR.

All of these can and should be fixed upstream, there is absolutely no reason to add a hook for this.

File: pkgs/build-support/setup-hooks/update-golang-sys-hook.sh
Comment (wegank) [reply pkgs/build-support/setup-hooks/update-golang-sys-hook.sh]: Done.

**Comment by avanov** [comment]
@zowoq should users of the currently failing derivations wait until the respective maintainers update their dependencies to support ARM, and then wait a little longer while nixpkgs maintainers provide respective PRs for every individual failing package? The issue with `sys` must be bothering all MacOS ARM users since November 2020 and two years later nixpkgs master still doesn't have it fixed for many Go packages. Upstream fixing can be happening in parallel to unblocking currently failing infra tooling needed on MacOS. The hook can safely be deprecated for updated upstream packages  when they get updated.

**Comment by zowoq** [comment]
> The issue with sys must be bothering all MacOS ARM users since November 2020 and two years later nixpkgs master still doesn't have it fixed for many Go packages

Can't be much of an issue if it's been a problem for two years and they still haven't been patched.

**Comment by avanov** [comment]
> Can't be much of an issue if it's been a problem for two years and they still haven't been patched.

Can't be much of an issue to you, but if there's an effort being put at the moment to fix it universally for many packages at once, it is definitely an issue for some. Why opposing this effort if your proposal could be worked on in parallel?

**Comment by zowoq** [comment]
> should users of the currently failing derivations wait until the respective maintainers update their dependencies to support ARM, and then wait a little longer while nixpkgs maintainers provide respective PRs for every individual failing package

Yes. 

> Upstream fixing can be happening in parallel to unblocking currently failing infra tooling needed on MacOS. 

No incentive for people to submit fixes upstream if we paper over it here.

**Comment by avanov** [comment]
> No incentive for people to submit fixes upstream if we paper over it here.

The actual incentive here is to compile a list of packages that DO NEED fixing of this kind, which enables anyone interested in propagating the required changes upstream to do it in bulk with the provided list of items. You could be that person too. All of that in parallel to actually fixing ARM environment for nixpkgs users. So, where's the downside?

**Comment by zowoq** [comment]
> The actual incentive here is to compile a list of packages that DO NEED fixing of this kind, which enables anyone interested in propagating the required changes upstream to do it in bulk with the provided list of items

This can be done without this PR.

> You could be that person too. 

lol, thanks. I think I've already been doing enough go maintenance in nixpkgs the last couple of years.

>  So, where's the downside?

The down side is the nixpkgs go maintainers getting stuck with a mess that needs to be cleaned up. Go packaging is bad enough as it is without adding another hack for a handful of packages.

**Comment by avanov** [comment]
> This can be done without this PR.

That's not an excuse to block this PR, as it has a valid use case.

> lol, thanks. I think I've already been doing enough go maintenance in nixpkgs the last couple of years.

There may be someone else, the point is that there's no need to wait for that someone else to productively unblock people downstream - the end users of `nixpkgs` the project exists for in the first place. Upstream fixing is a noble undertaking that may happen along the way, it doesn't have to block the main purpose of delivering `pkgs.<name>` to customers.

> The down side is the nixpkgs go maintainers getting stuck with a mess that needs to be cleaned up. 

A similar kind of situation is happening almost on every GHC Haskell release, but people do [create patches, overrides, and hooks](https://github.com/NixOS/nixpkgs/blob/ea469446d8ab8b3a951f3332d4c943d2773bd3f5/pkgs/development/haskell-modules/overrides.nix) to fix the majority of the available package set in bulk for a new compiler version. They also propagate changes to the upstream packages in parallel. With a decent amount of documentation and naming conventions it's even maintainable.

**Comment by zowoq** [comment]
> That's not an excuse to block this PR

Yes, it is.

>  There may be someone else

And if there isn't someone else it's going to become my problem.

> point is that there's no need to wait for that someone else to productively unblock people downstream

Again, already been waiting for two years.

>  Upstream fixing is a noble undertaking that may happen along the way

So this PR won't actually incentivize people to submit upstream fixes?

> A similar kind of situation is happening almost on every GHC Haskell release ...

This is a very different.

**Comment by avanov** [comment]
> Yes, it is.

lol, no

> And if there isn't someone else it's going to become my problem.

it is already your problem that packages are not supported on ARM, there's a working branch with fixes for nixpkgs specifically. Why would you predicate fixing nixpkgs derivations on ARM on the ability and desire of upstream owners to accept patches for all broken packages on a timely manner?

> Again, already been waiting for two years.

have you considered it could be because of your unwillingness to unblock potential maintainers from delivering fixes to end users so that they opt not to engage into longer term commitments for a voluntary work of this kind? It could be that adding `pkgs.<name>.overrieAttrs` on their side is easier than dealing with a situation like this one.

> So this PR won't actually incentivize people to submit upstream fixes?

It doesn't follow from the point I made about parallelisation. What is clear though, is that the fixes under this PR do fix packages for downstream users without blocking others to submit patches upstream.

> This is a very different.

In what ways precisely?

**Comment by zowoq** [comment]
>  Why would you predicate fixing nixpkgs derivations on ARM on the ability and desire of upstream owners to accept patches for all broken packages in a timely manner?

They don't need to accept them, they just need to be submitted upstream and then the patches can be `fetchpatch`ed  here.

IIRC this is the third or fourth "patch go.{mod,sum}" PR submitted in the last couple of years and it's always been the same answer. 

Not wasting any more time debating this.

@wegank Thank you, I appreciate the effort but this isn't going to be merged. Closing.

**Comment by avanov** [comment]
> They don't need to accept them, they just need to be submitted upstream and then the patches can be fetchpatched here.

that's not a definition of upstream submission as the submissions would only exist in forked instances of the upstream repositories with an unclear userbase and a lifetime.

> IIRC this is the third or fourth "patch go.{mod,sum}" PR submitted in the last couple of years and it's always been the same answer.

> Not wasting any more time debating this.

you haven't answered any of the questions on point, there was hardly a debate, only unsubstantiated opinions and hampering contributors.

**Comment by Ma27** [comment]
Cobsidering that the fix is relatively easynfor individual packages AFAIU: how about filing upstream patches and if people actually need the fixes, we can pull in the patches here?

**Comment by zowoq** [comment]
> Cobsidering that the fix is relatively easynfor individual packages AFAIU: how about filing upstream patches and if people actually need the fixes, we can pull in the patches here?

Not sure if I've misunderstood you or not, isn't this the same as what I said in a previous comment? 

> They don't need to accept them, they just need to be submitted upstream and then the patches can be `fetchpatch`ed here.

**Comment by SuperSandro2000** [comment]
> Cobsidering that the fix is relatively easynfor individual packages AFAIU: how about filing upstream patches and if people actually need the fixes, we can pull in the patches here?

I did this already for a handful of packages. Half of those did upstream submitted patch did not apply to the release version and I needed to redo the patch. gomod is not helping in this either and is more an annoyance with its strictness.

**Comment by SuperSandro2000** [comment]
I must say I like the idea of this hook but it modifies the vendor hash FOD which is already a big problem and sometimes silently out of date for months without anyone noticing it.

**Pull request closed without merging**
